File: mobile/lib/domain/merchant.dart
import 'package:offline_wallet/core/money.dart';

/// MerchantWallet — the two buckets a merchant cares about (FR-MER-02): funds
/// received but not yet settled, and funds already settled. Task 4: both start
/// at zero (settlement lands in a later task). Immutable.
class MerchantWallet {
  final Money pendingSettlement;
  final Money settled;

  const MerchantWallet({
    required this.pendingSettlement,
    required this.settled,
  });

  static MerchantWallet empty() =>
      MerchantWallet(pendingSettlement: Money.zero(), settled: Money.zero());

  Money get total => pendingSettlement.add(settled);

  @override
  bool operator ==(Object other) =>
      other is MerchantWallet &&
      other.pendingSettlement == pendingSettlement &&
      other.settled == settled;

  @override
  int get hashCode => Object.hash(pendingSettlement, settled);
}

/// Merchant — Merchant Mode entity (FR-MER-01). A role on an existing account:
/// minted with a Merchant ID and an empty wallet, no separate registration.
class Merchant {
  final String merchantId;
  final String accountId;
  final String displayName;
  final MerchantWallet wallet;

  const Merchant({
    required this.merchantId,
    required this.accountId,
    required this.displayName,
    required this.wallet,
  });

  @override
  bool operator ==(Object other) =>
      other is Merchant &&
      other.merchantId == merchantId &&
      other.accountId == accountId &&
      other.displayName == displayName &&
      other.wallet == wallet;

  @override
  int get hashCode => Object.hash(merchantId, accountId, displayName, wallet);
}

/// QrPayload — payment-QR contents (FR-PAY-01). Mirrors the wire contract
/// defined by PAYMENT_PROTOCOL.md §5: {v, typ, mid, n, ts, amt?}, carrying no
/// secrets. Task 4 is a placeholder: no signing, no single-use nonce tracking
/// yet.
class QrPayload {
  final int v;
  final String typ;
  final String merchantId;
  final String nonce;
  final int ts; // epoch seconds
  final int? amountPaise;

  const QrPayload({
    required this.v,
    this.typ = 'offer-req',
    required this.merchantId,
    required this.nonce,
    required this.ts,
    this.amountPaise,
  });

  /// True for a Fixed Amount Payment Request (Task 6.7): the payer never
  /// enters an amount. False is an Open Amount Payment Request: the payer
  /// enters the amount after scanning.
  bool get isFixedAmount => amountPaise != null;
}

File: backend/src/modules/merchant/domain/merchant.ts
import { Money } from '../../../shared/money';
import { randomUUID } from 'crypto';

/**
 * MerchantWallet — value object holding the two buckets a merchant cares about
 * (FR-MER-02): funds received but not yet settled, and funds already settled.
 * Task 4: both start at zero; settlement (a later task) will move value from
 * pendingSettlement → settled. Immutable.
 */
export class MerchantWallet {
  constructor(
    readonly pendingSettlement: Money,
    readonly settled: Money,
  ) {}

  static empty(): MerchantWallet {
    return new MerchantWallet(Money.zero(), Money.zero());
  }

  /** Total value the merchant has taken in (pending + settled). */
  get total(): Money {
    return this.pendingSettlement.add(this.settled);
  }
}

/**
 * Merchant — aggregate root for Merchant Mode (FR-MER-01, ARCHITECTURE.md §4.1).
 * A Merchant is a *role* on an existing account: enabling Merchant Mode mints a
 * Merchant ID and an empty merchant wallet with no separate registration.
 * Task 4 scope: no cryptography, no bank binding, no settlement.
 * Immutable: any future state change returns a new Merchant.
 */
export class Merchant {
  constructor(
    readonly merchantId: string,
    readonly accountId: string,
    readonly displayName: string,
    readonly wallet: MerchantWallet,
    readonly createdAt: Date,
  ) {}

  /**
   * Enable Merchant Mode for an account. Generates a fresh Merchant ID and an
   * empty merchant wallet.
   */
  static create(accountId: string, displayName: string, now: Date): Merchant {
    return new Merchant(
      generateMerchantId(),
      accountId,
      displayName,
      MerchantWallet.empty(),
      now,
    );
  }
}

/**
 * Merchant ID generation. Format: "MER-" + 12 uppercase hex chars derived from
 * a CSPRNG UUID. The ID is public and non-secret — it is embedded in QR
 * payloads (FR-PAY-01) — so it carries no key material.
 */
export function generateMerchantId(): string {
  const hex = randomUUID().replace(/-/g, '').slice(0, 12).toUpperCase();
  return `MER-${hex}`;
}
